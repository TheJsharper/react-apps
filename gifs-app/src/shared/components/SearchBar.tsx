import { useEffect, useState } from "react";

export interface SearchBarProps {
    placeholder: string;
    onSearchSubmit: (search: string) => void
}

export default function SearchBar({ placeholder, onSearchSubmit }: SearchBarProps) {
    const [searchTerm, setSearchTerm] = useState<string>('');

    const handleSubmit = (e: React.FormEvent) => {
        e.preventDefault();
        onSearchSubmit(searchTerm);
        console.log("Search submitted:", searchTerm);
        setSearchTerm('');
    };

    useEffect(() => {
        const timer = setTimeout(() => {
            console.log("Effect ran: ", searchTerm);
            onSearchSubmit(searchTerm);
        }, 700);

        return () => {
            clearTimeout(timer);
            console.log("Cleanup on unmount or before next effect");
        };
    }, [searchTerm, onSearchSubmit]);

    return (
        <form className="search-container" onSubmit={handleSubmit}>
            <input type="text" placeholder={placeholder} value={searchTerm} onChange={(e) => setSearchTerm(e.target.value)}
                onKeyDown={(e) => e.key === 'Enter' && handleSubmit(e)} />
            <button onClick={handleSubmit}> Search</button>
        </form>
    )
}

