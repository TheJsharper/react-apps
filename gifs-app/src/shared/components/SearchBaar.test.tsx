import { describe, expect, test, vi } from "vitest";
import SearchBar from "./SearchBar";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";

describe("SearchBaar", () => {
    test("should render correctly", () => {
        // Test implementation goes here
        const { container } = render(<SearchBar onSearchSubmit={() => { }} placeholder="Search..." />);

        screen.debug();

        expect(container).toMatchSnapshot();

        expect(screen.getByRole("textbox")).toBeDefined();

        expect(screen.getByRole("button")).toBeDefined();
    });

    test("should call onSearchSubmit with the correct value after 700ms debounce with setTimeout", async () => {
        const onSearchSubmit = vi.fn();

        render(<SearchBar onSearchSubmit={onSearchSubmit} placeholder="Search..." />);

        const input = screen.getByRole("textbox");

        fireEvent.change(input, { target: { value: 'test' } });

        screen.debug();
        // await new Promise((r) => setTimeout(r, 701));
        // vi.advanceTimersByTime(700);

        await waitFor(() => {
            expect(onSearchSubmit).toHaveBeenCalled();

            // Fast-forward time by 700ms
            expect(onSearchSubmit).toHaveBeenCalledWith("test");
        }, { timeout: 20000 });
    });



    test("should call just once  with the last value (debounce) after multiple changes with setTimeout", async () => {
        const onSearchSubmit = vi.fn();
        render(<SearchBar onSearchSubmit={onSearchSubmit} placeholder="Search..." />);

        const input = screen.getByRole("textbox");
        fireEvent.change(input, { target: { value: 't' } });

        fireEvent.change(input, { target: { value: 'te' } });
        fireEvent.change(input, { target: { value: 'tes' } });
        fireEvent.change(input, { target: { value: 'test' } });
        screen.debug();
        // await new Promise((r) => setTimeout(r, 701));
        await waitFor(() => {
            expect(onSearchSubmit).toHaveBeenCalledTimes(1);
            expect(onSearchSubmit).toHaveBeenCalledWith("test");
        });
        // vi.advanceTimersByTime(700);
    });

    test("should call onSearchSubmit on form submit and clear input", async () => {
        const onSearchSubmit = vi.fn();
        render(<SearchBar onSearchSubmit={onSearchSubmit} placeholder="Search..." />);

        const input = screen.getByRole("textbox") as HTMLInputElement;
        fireEvent.change(input, { target: { value: 'test' } });

        const button = screen.getByRole("button");
        fireEvent.click(button);

        await waitFor(() => {
            expect(onSearchSubmit).toHaveBeenCalledTimes(1);
            expect(onSearchSubmit).toHaveBeenCalledWith("test");
            expect(input.value).toBe("");
        });
    });

   test("should have  the input  correctly  placeholder", async () => {
        const placeholderText = "Search for a gif...";

        render(<SearchBar onSearchSubmit={() => { }} placeholder={placeholderText} />);
        
        screen.debug();

        const input = screen.getByPlaceholderText(placeholderText);
        
        expect(input).toBeDefined();
    });

    // Add more tests as needed
});