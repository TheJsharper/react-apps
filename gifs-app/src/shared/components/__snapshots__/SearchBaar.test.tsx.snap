// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBaar > should render correctly 1`] = `
<div>
  <form
    class="search-container"
  >
    <input
      placeholder="Search..."
      type="text"
      value=""
    />
    <button>
       Search
    </button>
  </form>
</div>
`;
